Fix WEBGL_draw_buffers failures

lexaknyazev opened the pull request:
#### dccb2b202e36836723cbaa78fc6186503cc22c04
<pre>
Fix WEBGL_draw_buffers failures
<a href="https://bugs.webkit.org/show_bug.cgi?id=242028">https://bugs.webkit.org/show_bug.cgi?id=242028</a>

Reviewed by Kimmo Kinnunen.

The drawBuffersWEBGL must generate INVALID_OPERATION for invalid
arguments, unless the array length is greater than the number of
draw buffers.

When the default backbuffer is set to NONE, switch it to BACK,
or actually to COLOR_ATTACHMENT0 because it is simulated, before
doing implicit clears and restore to NONE afterwards.

* LayoutTests/TestExpectations:
* LayoutTests/fast/canvas/webgl/webgl-draw-buffers-expected.txt: Removed.
* LayoutTests/fast/canvas/webgl/webgl-draw-buffers.html: Removed.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* LayoutTests/webgl/1.0.3/conformance/extensions/webgl-draw-buffers-expected.txt:
* LayoutTests/webgl/2.0.0/conformance/extensions/webgl-draw-buffers-expected.txt:
* Source/WebCore/html/canvas/WebGLDrawBuffers.cpp:
(WebCore::WebGLDrawBuffers::drawBuffersWEBGL):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::ScopedEnableBackbuffer::ScopedEnableBackbuffer):
(WebCore::ScopedEnableBackbuffer::~ScopedEnableBackbuffer):
(WebCore::WebGLRenderingContextBase::clearIfComposited):
(WebCore::WebGLRenderingContextBase::setBackDrawBuffer):

Canonical link: <a href="https://commits.webkit.org/251910@main">https://commits.webkit.org/251910@main</a>
</pre>

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -512,6 +512,39 @@ class ScopedDisableRasterizerDiscard {
+public:
+    explicit ScopedEnableBackbuffer(GraphicsContextGL& context, GCGLenum backDrawBuffer, bool isWebGL2)
+        : m_context(context)
+        , m_backDrawBuffer(backDrawBuffer)
```
This is really a boolean variable, so you can consider changing this to `m_didChangeBackbuffer` or similar.

kkinnunen-apple replied:
that `ASSERT(backDrawBuffer == GraphicsContextGL::NONE || backDrawBuffer == GraphicsContextGL::COLOR_ATTACHMENT0);` would perhaps clarify the code.

kkinnunen-apple replied:
`ASSERT(backDrawBuffer == GraphicsContextGL::NONE || backDrawBuffer == GraphicsContextGL::BACK)`

lexaknyazev committed:
Fix WEBGL_draw_buffers failures
https://bugs.webkit.org/show_bug.cgi?id=242028

Reviewed by Kimmo Kinnunen.

The drawBuffersWEBGL must generate INVALID_OPERATION for invalid
arguments, unless the array length is greater than the number of
draw buffers.

When the default backbuffer is set to NONE, switch it to BACK,
or actually to COLOR_ATTACHMENT0 because it is simulated, before
doing implicit clears and restore to NONE afterwards.

* LayoutTests/TestExpectations:
* LayoutTests/fast/canvas/webgl/webgl-draw-buffers-expected.txt: Removed.
* LayoutTests/fast/canvas/webgl/webgl-draw-buffers.html: Removed.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* LayoutTests/webgl/1.0.3/conformance/extensions/webgl-draw-buffers-expected.txt:
* LayoutTests/webgl/2.0.0/conformance/extensions/webgl-draw-buffers-expected.txt:
* Source/WebCore/html/canvas/WebGLDrawBuffers.cpp:
(WebCore::WebGLDrawBuffers::drawBuffersWEBGL):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::ScopedEnableBackbuffer::ScopedEnableBackbuffer):
(WebCore::ScopedEnableBackbuffer::~ScopedEnableBackbuffer):
(WebCore::WebGLRenderingContextBase::clearIfComposited):
(WebCore::WebGLRenderingContextBase::setBackDrawBuffer):

Canonical link: https://commits.webkit.org/251910@main
